cmd/uplink/cmd/share.go

```diff
@@ -0,0 +1,147 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package cmd
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/araddon/dateparse"
+	"github.com/spf13/cobra"
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/internal/fpath"
+	libuplink "storj.io/storj/lib/uplink"
+	"storj.io/storj/pkg/cfgstruct"
+	"storj.io/storj/pkg/encryption"
+	"storj.io/storj/pkg/macaroon"
+	"storj.io/storj/pkg/process"
+)
+
+var shareCfg struct {
+	DisallowReads     bool     `default:"false" help:"if true, disallow reads"`
+	DisallowWrites    bool     `default:"false" help:"if true, disallow writes"`
+	DisallowLists     bool     `default:"false" help:"if true, disallow lists"`
+	DisallowDeletes   bool     `default:"false" help:"if true, disallow deletes"`
+	Readonly          bool     `default:"false" help:"implies disallow_writes and disallow_deletes"`
+	Writeonly         bool     `default:"false" help:"implies disallow_reads and disallow_lists"`
+	NotBefore         string   `help:"disallow access before this time"`
+	NotAfter          string   `help:"disallow access after this time"`
+	AllowedPathPrefix []string `help:"whitelist of bucket path prefixes to require"`
+}
+
+func init() {
+	// sadly, we have to use addCmd so that it adds the cfg struct to the flags
+	// so that we can open projects and buckets. that pulls in so many unnecessary
+	// flags which makes figuring out the share command really hard. oh well.
+	shareCmd := addCmd(&cobra.Command{
+		Use:   "share",
+		Short: "Creates a possibly restricted api key",
+		RunE:  shareMain,
+	}, RootCmd)
+
+	cfgstruct.Bind(shareCmd.Flags(), &shareCfg)
+}
+
+func parseHumanDate(date string, now time.Time) (*time.Time, error) {
+	if date == "" {
+		return nil, nil
+	} else if date == "now" {
+		return &now, nil
+	} else if date[0] == '+' {
+		d, err := time.ParseDuration(date[1:])
+		t := now.Add(d)
+		return &t, errs.Wrap(err)
+	} else if date[0] == '-' {
+		d, err := time.ParseDuration(date[1:])
+		t := now.Add(-d)
+		return &t, errs.Wrap(err)
+	} else {
+		t, err := dateparse.ParseAny(date)
```

Comment:
My concern with this is that it picks an arbitrary resolution for dates in the form of `01.03.2019` and `01/03/2019` rather than using locale. I would rather restrict us to ISO8601 like formats (year, month, day etc.).

Reply:
or return an error when the date is ambiguous using https://godoc.org/github.com/araddon/dateparse#ParseStrict

Reply:
Would it be better if it printed out the time in an unambiguous format if specified? Personally, I can never remember date formats.

Reply:
For me being stuck on one format is OK, but this a user level application, and you know that can make some users uncomfortable.

Reply:
I wouldn't call `uplink` really an user level application. I can see it only being used by programmers and scripting. Programmers can probably figure out ISO8601. Scripters probably want something consistent that doesn't break when they move code from one source to another.

Reply:
good to me, my sentence should have been

> For me being stuck on one single format is preferable (and ISO8601 is the best) ...

Reply:
Alright, sounds like the consensus is to just require ISO8601.